### ISSUE TITLE
IDE0059 does not work for variables in the presence of an expression tree

### ISSUE THREAD

**Issue by j2jensen** [dotnet/roslyn#80033]
**Version Used**: .NET SDK 9.0.302

**Steps to Reproduce**:

1. In a .NET project, set IDE0059 to be active (e.g. warning level)
2. Create a method that stores a (non-constant) value into a variable, but doesn't use it. Verify that IDE0059 complains.
    ```
    public class A
    {
        public void Foo()
        {
            var a = Random.Shared.Next();   // unnecessary assignment of a value to 'a'
        }
    }
    ```
3. Now introduce an Expression Tree into the method.
    ```
    public class A
    {
        public void Foo()
        {
            var a = Random.Shared.Next();
            Expression<Func<int>> _ = () => 1;
        }
    }
    ```

**Diagnostic Id**: IDE0059: Remove unnecessary value assignment

**Expected Behavior**:
IDE0059 should continue warning about the unused variable.

**Actual Behavior**:
No warning appears.

**Note**:
If the variable is set to a constant or a literal, IDE0059 continues to see it as a problem. For example `var a = "";` will get matched, but `var a = string.Empty;` will not.

If it turns out to be very difficult to fix this, due to some quirk of Expression Trees in the compiler, I would understand. But perhaps that should be documented so people don't spend as much time trying to figure out why they're getting this warning in some methods and not in others.

**Comment by MihaZupan** [comment]
Possibly a dupe of #38376

**Comment by j2jensen** [comment]
> Possibly a dupe of [#38376](https://github.com/dotnet/roslyn/issues/38376)

Yes, it could be. Interesting to note that if you just use `Func<int>` instead of `<Expression<Func<int>>` in my example, the bug does not happen. But passing the Func as an argument as in the example in #38376 does cause the same behavior.

**Comment by CyrusNajmabadi** [comment]
Looks intentional: https://github.com/dotnet/roslyn/pull/31775

**Comment by CyrusNajmabadi** [comment]
@mavasani Do you remember why ExprTree isn't supported here?  Is it a case of IOp not representing things?  Something else?  Thanks!

**Comment by mavasani** [comment]
Both - IOp not supporting it and complexity in the analyzer itself.. we had decided to just put the work on hold that time

**Comment by CyrusNajmabadi** [comment]
Clsoing out as this is out of scope given the limitations of ExprTrees and IOp.
